### PR TITLE
Improve `finite_diff` for custom data type implementations and fix corresponding test

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
An  existing test in the test suite for `gradients.finite_diff` had a typo, turning it into a copy of a test for `param_shift` instead, see #3398.
In addition, `finite_diff` requires unnecessarily many magic methods to be implemented for a custom data type.

**Description of the Change:**
The typo is fixed. In addition, `finite_diff` is changed so that fewer elementary/magic methods need to be implemented when using it with a custom data type. In particular, `finite_diff` previously required `__truediv__`, `__radd__` and `__rmul__` to be defined (among others), which is not longer the case.

**Benefits:**
More convenient usage of `finite_diff` with custom data types. Fixed test to actually cover what we want to test.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
#3398

### CODE REVIEW

File: pennylane/gradients/finite_difference.py [hunk 1]
```diff
@@ -724,13 +724,13 @@ def processing_fn(results):
 
             # compute the linear combination of results and coefficients
             res = qml.math.stack(res)
-            g = sum(c * r for c, r in zip(coeffs, res))
+            g = sum(r * c for c, r in zip(coeffs, res))
```
Comment (dwierichs) [inline pennylane/gradients/finite_difference.py:727]: This makes it possible to use a data type for the results that does not define `__rmul__`

File: pennylane/gradients/finite_difference.py [hunk 2]
```diff
@@ -724,13 +724,13 @@ def processing_fn(results):
 
             # compute the linear combination of results and coefficients
             res = qml.math.stack(res)
-            g = sum(c * r for c, r in zip(coeffs, res))
+            g = sum(r * c for c, r in zip(coeffs, res))
 
             if c0 is not None:
                 # add on the unshifted term
-                g = g + c0 * r0
+                g = g + r0 * c0
```
Comment (dwierichs) [inline pennylane/gradients/finite_difference.py:731]: This makes it possible to use a data type for the results that does not define `__rmul__`

File: pennylane/gradients/finite_difference.py [hunk 3]
```diff
@@ -724,13 +724,13 @@ def processing_fn(results):
...
-            g = sum(c * r for c, r in zip(coeffs, res))
+            g = sum(r * c for c, r in zip(coeffs, res))
 
             if c0 is not None:
                 # add on the unshifted term
-                g = g + c0 * r0
+                g = g + r0 * c0
 
-            grads.append(g / (h**n))
+            grads.append(g * (h ** (-n)))
```
Comment (dwierichs) [inline pennylane/gradients/finite_difference.py:733]: This removes the necessity to implement `__truediv__` for the data type, the standard `/` operation between the custom object and a standard dtype like `float`.

File: tests/gradients/test_finite_difference.py [hunk 4]
```diff
@@ -373,9 +373,12 @@ def __mul__(self, other):
                 return SpecialObject(self.val * other)
 
             def __add__(self, other):
-                new = self.val + other.val if isinstance(other, self.__class__) else other
+                new = self.val + (other.val if isinstance(other, self.__class__) else other)
```
Comment (dwierichs) [inline tests/gradients/test_finite_difference.py:376]: This was a silent bug: The addition previously was executed _before_ the Boolean clause, yielding `new=other` for the case of `not isinstance(other, self.__class__)`.

File: tests/gradients/test_parameter_shift.py [hunk 5]
```diff
@@ -1297,7 +1297,7 @@ def __mul__(self, other):
                 return SpecialObject(self.val * other)
 
             def __add__(self, other):
-                newval = self.val + other.val if isinstance(other, self.__class__) else other
+                newval = self.val + (other.val if isinstance(other, self.__class__) else other)
```
Comment (dwierichs) [inline tests/gradients/test_parameter_shift.py:1300]: This was a silent bug: The addition previously was executed _before_ the Boolean clause, yielding `new=other` for the case of `not isinstance(other, self.__class__)`.

File: tests/gradients/test_finite_difference.py [hunk 6]
```diff
@@ -392,7 +395,7 @@ class DeviceSupportingSpecialObservable(DefaultQubit):
 
             @staticmethod
             def _asarray(arr, dtype=None):
-                return arr
+                return np.asarray(arr)
```
Comment (dwierichs) [inline tests/gradients/test_finite_difference.py:398]: Pennylane's, ie. Autograd's, NumPy arrays behave _much_ more reasonable when they are zero-dimensional, therefore a custom implementation should use those, not vanilla numpy arrays.
An example is that `np.array(MyCoolObject(0.4), dtype=object) * 0.3` returns an array with Pennylane's NumPy, but a `MyCoolObject` instance for vanilla NumPy (provided that `__mul__` is implemented and returns a `MyCoolObject` instance itself).

File: tests/gradients/test_finite_difference.py [hunk 7]
```diff
@@ -407,12 +410,12 @@ def expval(self, observable, **kwargs):
 
         dev = DeviceSupportingSpecialObservable(wires=1, shots=None)
 
-        @qml.qnode(dev, diff_method="parameter-shift")
+        @qml.qnode(dev, diff_method="finite-diff")
```
Comment (dwierichs) [inline tests/gradients/test_finite_difference.py:413]: This is the typo that led to this PR

File: tests/gradients/test_finite_difference.py [hunk 8]
```diff
@@ -407,12 +410,12 @@ def expval(self, observable, **kwargs):
...
         dev = DeviceSupportingSpecialObservable(wires=1, shots=None)
 
-        @qml.qnode(dev, diff_method="parameter-shift")
+        @qml.qnode(dev, diff_method="finite-diff")
         def qnode(x):
             qml.RY(x, wires=0)
             return qml.expval(SpecialObservable(wires=0))
 
-        @qml.qnode(dev, diff_method="parameter-shift")
+        @qml.qnode(dev, diff_method="finite-diff")
```
Comment (dwierichs) [inline tests/gradients/test_finite_difference.py:418]: To check up to good numerical precision, we should choose the same method here as above.

File: pennylane/gradients/finite_difference.py (hunk 3 above)
Comment (rmoyard) [reply pennylane/gradients/finite_difference.py:733]: Nice hack 👍

**Review by rmoyard — COMMENTED** [review]
Thanks @dwierichs for solving that and correcting the tests. Could you also correct the tests in the new return type folder: `test_finite_difference_new` and `test_parameter_shift_new`? Otherwise it looks good 💯

**Comment by dwierichs** [comment]
> Could you also correct the tests in the new return type folder: test_finite_difference_new and test_parameter_shift_new?

Done :heavy_check_mark:

**Comment by rmoyard** [comment]
I think that it is a mistake from our side but can you uncomment the tests for the new return type?

**Comment by dwierichs** [comment]
Are the #TODO's still up to date or should I just delete those?

**Comment by rmoyard** [comment]
I think we should remove them, because the integration is done.

**Review by rmoyard — APPROVED** [review]
Thanks for fixing that, it looks good!

**Pull request merged**